design/app-to-app.md

```diff
@@ -0,0 +1,111 @@
+// https://github.com/microsoft/mixed-reality-extension-sdk/blob/master/packages/sdk/src/core/context.ts
+
+/**
+ * Registers a set of callbacks invoked at various points in the processing
+ * of a sharing message. Only one callback may be registered at a time
+ * for a given callback type.
+ * 
+ * The messageReceived callback must be defined, but the isMessageDesired
+ * callback is optional.
+ * 
+ * Returns true if the callback was successfully registered.
+ * @event
+ */
+public onSharingMessage(messageType: string, callbacks: SharingCallbacks): bool {
+    // [... Register listener ...]
+    return true; // If successful
+}
+
+/**
+ * Removes the onSharingMessage callback, if it is registered.
+ * 
+ * Returns true if the callback was successfully unregistered.
+ * @event
+ */
+public offSharingMessage(messageType: string, callbacks: SharingCallbacks): bool {
+    // [... Deregister listener ...]
+    return true; // If successful
+}
+```
+
+## Implementation Suggestions
+
+1. Modify `context` to expose callback registration mechanism
+2. Adjust `contextInternal` and perhaps `multiPeerAdapter` to message the desire for each MessageType callback to each client (and corresponding remove messages). For callbacks registered for all sessions or all users, this my be sent after connect
+3. Implement the "Photo" message type. Send and receive encoded photos via the existing websocket used by multiPeerAdapter. Future message types (e.g., streaming video) may need to establish related side-channels.
+
+### Sample data flow for photo sharing
+1. On connect (or on subsequent sharing message registration), MRE app sends a `SharingRegistration` json payload consisting of
+    * `messageType` - `string` (as Described in [API Design](#API-Design))
+2. MRE Client sends a `SharingMessage` json payload, consiting of the following fields (as Described in [API Design](#API-Design))
+    * `messageType` - `string` 
+    * `photoType` - `string`
+    * `dataSize` - `number`
+    * `transmissionMethod` - `string`, presently set to the value "websocketbinary", but may be extended, for example, to support side channel for future message types, or alternative trasnport for existing message types.
+3. MRE app invokes `SharingCallbacks.isMessageDesired` callback if present to determine whether the message is desired. If a `SharingCallback` is registered for the given `messageType` but no `isMessageDesired` is present, proceed as if the message IS desired. If no `SharingCallback` is registered for the given `messageType`, proceed as if the message is NOT desired.
+4. MRE app sends a `SharingMessageResponse` json payload, consiting of the following fields:
+    * `result` - `boolean` true if the message is desired, false otherwise.
+5. If the message was desired, the MRE host should transmit the bytes of the image immediately following the `SharingMessageResponse` in binary format as per https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send
```

Comment:
I think we'll want adequate header information in the blob message to indicate that this is image data for a specific SharingMessage.  If we decided to add a new blob message in the future for a different purpose, there should be no ambiguity as to what a given blob is for.

If putting a header _in_ the blob feels excessive, then consider how to build a generic mechanism into the protocol for associating blob data with a JSON message.  I'd recommend thinking about that as a transport problem, absent semantics.,   That is, you should be able as much as possible to parse the protocol and route messages with no with no knowledge of particular message types or exchanges.  This becomes important to extensibility & back-compat.

Reply:
i'll have us resend the message (in text form) on the binary side channel. lemme know if my latest commit helps.